docs/agents/agent-development.md

```diff
@@ -237,14 +237,33 @@ Global labels can be specified via the environment variable `ELASTIC_APM_GLOBAL_
 
 ### Transactions
 
+#### Transaction outcome
+
+The `outcome` property denotes whether the transaction represents a success or a failure from the perspective of the entity that produced the event.
+The APM Server converts this to the [`event.outcome`](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-outcome.html) field.
+This property is optional to remain backwards compatibility.
+If an agent doesn't report the `outcome` (or reports `null`), the APM Server sets the outcome to `"unknown"`.
```

Comment:
This means we have no way of leaving off the outcome, for cases where it does not make logical sense, per https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-outcome.html#ecs-event-outcome-unknown. What's the downside of just not setting the field if the agent doesn't report it?

What about agents start defaulting to setting the field to "unknown", and allow it to be unset in case it doesn't make sense - and the server will just ignore any empty/null outcome?

Reply:
Just saw your commit message:

> An outcome always makes sense for span or transaction events but sometimes it can't be determined.

Maybe I'm being too cautious but this assertion makes me uneasy, and I'm still wondering what we're gaining by making unknown equivalent to null.

Reply:
We'd always have a non-null outcome when using a recent APM Server which might be useful when connecting old agents to it. As the old agents don't send an outcome, it would mean that an outcome does not make sense for these events vs the outcome being unknown.

> Maybe I'm being too cautious but this assertion makes me uneasy

I might be wrong but I think spans and transactions are just a type of event where it always makes sense that they have an outcome. That is in contrast to log events, for example. Those types of events just don't have an outcome.

Do you have some specific scenarios in mind where it's not appropriate for a span or transaction to have an outcome?

@webmat how do you see this? Does it make sense to differentiate between `null` and `"unknown"` for the `event.outcome` of transactions and spans?

Reply:
> I think detecting older agents is better done via other information that's present in the events 

I'd argue that all old agents can't or don't determine the outcome, therefore `"unknown"` seems appropriate for them.

> On the other hand, if all of your events would be expected to have an outcome, but some simply can't be determined by the agent, then I agree it's fine to set the outcome to "unknown".

This proposal is just about adding an outcome to transactions and spans. They basically describe an event like handling an incoming or outgoing HTTP request or a method execution. This includes the duration of the event, the name, and more metadata. It's one event that covers the whole length of the operation i.e. it's not split into two for entry and exit of an operation. I can't think of a case where an `outcome` would be inappropriate for spans and transactions. 

APM defines more events but we don't plan to add outcome to metrics or error events.

Reply:
I'm probably being overcautious. I agree with setting to `unknown` by default (i.e. if omitted/null) for transactions and spans.